widget/list.go

```diff
@@ -0,0 +1,415 @@
+}
+
+type listItemRenderer struct {
+	*widget.ShadowingRenderer
+
+	item   *listItem
+	layout fyne.Layout
+}
+
+// MinSize calculates the minimum size of a listItem.
+// This is based on the size of the status indicator and the size of the child object.
+func (li *listItemRenderer) MinSize() (size fyne.Size) {
+	li.item.statusIndicator.SetMinSize(fyne.NewSize(theme.Padding(), li.item.child.Size().Height))
+	size = fyne.NewSize(theme.Padding()+li.item.child.Size().Width, li.item.child.Size().Height)
+	return
+}
+
+// Layout the components of the button widget
+func (li *listItemRenderer) Layout(size fyne.Size) {
+	li.LayoutShadow(size, fyne.NewPos(0, 0))
+	li.layout.Layout([]fyne.CanvasObject{li.item.statusIndicator, li.item.child}, size)
+}
+
+func (li *listItemRenderer) BackgroundColor() color.Color {
+	if li.item.background == nil {
+		return theme.BackgroundColor()
+	}
+
+	return li.item.background
+}
+
+func (li *listItemRenderer) Refresh() {
+	if li.item.selected {
+		li.item.statusIndicator.FillColor = theme.FocusColor()
+		canvas.Refresh(li.item.super())
+		return
+	}
+	if li.item.hovered {
+		li.item.statusIndicator.FillColor = theme.HoverColor()
+	} else {
+		li.item.statusIndicator.FillColor = theme.BackgroundColor()
+	}
+	canvas.Refresh(li.item.super())
+}
+
+type listLayout struct {
+	list      *List
+	itemCount int
+}
+
+var _ fyne.Layout = (*listLayout)(nil)
+
+func newListLayout(list *List) fyne.Layout {
+	return &listLayout{list: list}
+}
+
+func (l *listLayout) Layout(objects []fyne.CanvasObject, size fyne.Size) {
+	if l.list.offsetY != 0 {
+		return
+	}
+	y := 0
+	for _, child := range objects {
+		child.Move(fyne.NewPos(0, y))
+		y += l.list.itemMin.Height
+		child.Resize(fyne.NewSize(l.list.size.Width, l.list.itemMin.Height))
+	}
+}
+
+func (l *listLayout) MinSize(objects []fyne.CanvasObject) fyne.Size {
+
+	return fyne.NewSize(l.list.itemMin.Width+theme.Padding(),
+		l.list.itemMin.Height*l.list.Length())
+}
+
+func (l *listLayout) itemAppended(objects []fyne.CanvasObject) {
+	if len(objects) > 1 {
+		objects[len(objects)-1].Move(fyne.NewPos(0, objects[len(objects)-2].Position().Y+l.list.itemMin.Height))
+	} else {
+		objects[len(objects)-1].Move(fyne.NewPos(0, 0))
+	}
+	objects[len(objects)-1].Resize(fyne.NewSize(l.list.size.Width, l.list.itemMin.Height))
+}
+
+func (l *listLayout) itemPrepended(objects []fyne.CanvasObject) {
+	objects[0].Move(fyne.NewPos(0, objects[1].Position().Y-l.list.itemMin.Height))
+	objects[0].Resize(fyne.NewSize(l.list.size.Width, l.list.itemMin.Height))
+}
+
+type pool interface {
+	Count() int
+	Obtain() fyne.CanvasObject
+	Release(fyne.CanvasObject)
+}
+
+type listPool struct {
+	contents []fyne.CanvasObject
+}
+
+// Count returns the number of available items in the cache
+func (lc *listPool) Count() int {
+	return len(lc.contents)
+}
+
+// Obtain returns an item from the cache for use
+func (lc *listPool) Obtain() fyne.CanvasObject {
+	cacheLength := len(lc.contents)
+	if cacheLength == 0 {
```

Comment:
This variable isn't needed along with the check below...
If you check len(lc.contents) and return nil then you don't need to also check it later.

Reply:
agree

Reply:
This should now be fixed